### PR TITLE
title: virtualcluster join controller

### LINKED ISSUE #423: Add join controller to create cluster in virtual control plane
**What would you like to be added**:
Add join controller to create cluster in virtual control plane
proposal: https://kosmos-io.github.io/website/v0.2.0/proposals/k8s-in-k8s

**Why is this needed**:

### PR DESCRIPTION
**Pull request by GreatLazyMan**
#### What type of PR is this?
Add join controller to create cluster in virtual control plane
/kind feature

#### What does this PR do?
Add join controller to create cluster in virtual control plane

#### Which issue(s) does this PR fix?
Fixes #423
#### Does this PR introduce a user-facing change?

```

### CODE REVIEW

File: pkg/treeoperator/controller/virtualcluster_join_controller.go [hunk 1]
```diff
@@ -1,16 +1,452 @@
...
+		}
+	}
+	klog.Infof("Cluster %s has been deleted.", clusterName)
+	if err := c.RemoveFinalizer(ctx, vc); err != nil {
+		return fmt.Errorf("remove finalizer error: %v", err)
+	}
+	return nil
+}
+
+func (c *VirtualClusterJoinController) InitTargetKubeclinet(kubeconfigStream []byte) (error, versioned.Interface, kubernetes.Interface, extensionsclient.Interface) {
```
Comment (yuleichun-striving) [inline pkg/treeoperator/controller/virtualcluster_join_controller.go]: There are misspellings of words ｜ InitTargetKubeClient

File: pkg/treeoperator/controller/virtualcluster_join_controller.go [hunk 2]
```diff
@@ -1,16 +1,452 @@
...
+	if err != nil {
+		return err
+	}
+	err = c.Create(ctx, clustertreeSecret)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("install clustertree error, secret created failed: %v", err)
+		}
+	}
+	klog.Info("Secret has been created. ")
```
Comment (yuleichun-striving) [inline pkg/treeoperator/controller/virtualcluster_join_controller.go]: The log can be more detailed: please indicate which component the secret belongs to.

File: pkg/treeoperator/constants/constant.go [hunk 3]
```diff
@@ -3,4 +3,9 @@ package constants
 const (
 	InitControllerName = "virtual-cluster-init-controller"
 	JoinControllerName = "virtual-cluster-join-controller"
+
+	DefauleImageRepositoryEnv     = "IMAGE_REPOSITIRY"
+	DefauleImageVersionEnv        = "IMAGE_VERSION"
+	VirtualClusterStatusCompleted = "Completed"
+	VirtualClusterSFinalizerName  = "kosmos.io/virtual-cluster-finalizer"
```
Comment (yuleichun-striving) [inline pkg/treeoperator/constants/constant.go]: There is an extra character 'S' in the word ｜ VirtualClusterFinalizerName

File: pkg/treeoperator/controller/virtualcluster_join_controller.go [hunk 4]
```diff
@@ -1,16 +1,452 @@
...
+	_, err := kosmosClient.KosmosV1alpha1().Clusters().Update(context.TODO(), cluster, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Errorf("cluster %s failed remove finalizer: %v", cluster.Name, err)
+		return err
+	}
+	klog.Infof("update cluster after remove finalizer for cluster %s", cluster.Name)
+	return nil
+}
+
+func (c *VirtualClusterJoinController) UninstallClustertree(ctx context.Context, request reconcile.Request, vc *v1alpha1.VirtualCluster) error {
```
Comment (yuleichun-striving) [inline pkg/treeoperator/controller/virtualcluster_join_controller.go]: The first character T of tree needs to be capitalized ｜ ClusterTree

File: pkg/treeoperator/controller/virtualcluster_join_controller.go [hunk 5]
```diff
@@ -1,16 +1,452 @@
...
+	klog.Info("Start deleting kosmos-clustertree Deployment...")
+	clustertreeDeploy, err := kosmosctl.GenerateDeployment(treemanifest.ClusterTreeClusterManagerDeployment, treemanifest.DeploymentReplace{
+		Namespace:       request.Namespace,
+		ImageRepository: "null",
+		Version:         "null",
```
Comment (yuleichun-striving) [inline pkg/treeoperator/controller/virtualcluster_join_controller.go:71]: ImageRepostity value is null ?

File: pkg/treeoperator/controller/virtualcluster_join_controller.go [hunk 6]
```diff
@@ -1,16 +1,452 @@
...
+	return nil, kosmosClient, k8sClient, k8sExtensionsClient
+}
+
+func (c *VirtualClusterJoinController) DeployKosmos(ctx context.Context, request reconcile.Request, vc *v1alpha1.VirtualCluster) error {
+	klog.Infof("Start creating kosmos-clustertree secret")
+	clustertreeSecret, err := kosmosctl.GenerateSecret(treemanifest.ClusterTreeClusterManagerSecret, treemanifest.SecretReplace{
+		Namespace:  request.Namespace,
+		Cert:       cert.GetCrtEncode(),
+		Key:        cert.GetKeyEncode(),
+		Kubeconfig: vc.Spec.Kubeconfig,
```
Comment (yuleichun-striving) [inline pkg/treeoperator/controller/virtualcluster_join_controller.go:193]: The kubeconfig require base64 decryption?

File: pkg/treeoperator/controller/virtualcluster_join_controller.go [hunk 7]
```diff
@@ -1,16 +1,452 @@
...
+	var vc v1alpha1.VirtualCluster
+
+	if err := c.Get(ctx, request.NamespacedName, &vc); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		klog.Errorf("get %s error: %v", request.NamespacedName, err)
+		return reconcile.Result{RequeueAfter: utils.DefaultRequeueTime}, nil
+	}
+	if vc.DeletionTimestamp.IsZero() {
```
Comment (yuleichun-striving) [inline pkg/treeoperator/controller/virtualcluster_join_controller.go:391]: It is wrong to install services only based on whether vs is deleted or not.

File: pkg/treeoperator/controller/virtualcluster_join_controller.go (hunk 1 above)
Comment (GreatLazyMan) [reply pkg/treeoperator/controller/virtualcluster_join_controller.go]: fix it

File: pkg/treeoperator/controller/virtualcluster_join_controller.go (hunk 2 above)
Comment (GreatLazyMan) [reply pkg/treeoperator/controller/virtualcluster_join_controller.go]: add some log info

File: pkg/treeoperator/constants/constant.go (hunk 3 above)
Comment (GreatLazyMan) [reply pkg/treeoperator/constants/constant.go]: fix it

File: pkg/treeoperator/controller/virtualcluster_join_controller.go (hunk 4 above)
Comment (GreatLazyMan) [reply pkg/treeoperator/controller/virtualcluster_join_controller.go]: fix it

File: pkg/treeoperator/controller/virtualcluster_join_controller.go (hunk 5 above)
Comment (GreatLazyMan) [reply pkg/treeoperator/controller/virtualcluster_join_controller.go:71]: it's a delete action, test pass

File: pkg/treeoperator/controller/virtualcluster_join_controller.go (hunk 6 above)
Comment (GreatLazyMan) [reply pkg/treeoperator/controller/virtualcluster_join_controller.go:193]: it's a kubeconfig in secret, no need to do this

File: pkg/treeoperator/controller/virtualcluster_join_controller.go (hunk 7 above)
Comment (GreatLazyMan) [reply pkg/treeoperator/controller/virtualcluster_join_controller.go:391]: fix it

**Comment by duanmengkk** [comment]
rebase commits

**Comment by GreatLazyMan** [comment]
> rebase commits

rebased

**Pull request merged**